flips/network-api.md

```diff
@@ -0,0 +1,93 @@
+:-------------- |:--------------------------------------------------------- |
+| **FLIP #**    | [1306](https://github.com/onflow/flow-go/pull/1306) |
+| **Author(s)** | Simon Zhu (simon.zhu@dapperlabs.com)                      |
+| **Sponsor**   | Simon Zhu (simon.zhu@dapperlabs.com)                      |
+| **Updated**   | 9/16/2021                                                 |
+
+## Objective
+
+Refactor the networking layer to split it into separate APIs for the public and private network, allow us to implement a strict separation in the code between these two networks.
+
+Enable registering a custom message ID function for the gossip layer.
+
+## Current Implementation
+
+When the network layer receives a message, it will pass the message to the [`Engine`](https://github.com/onflow/flow-go/blob/7763000ba5724bb03f522380e513b784b4597d46/network/engine.go) registered on
+the corresponding channel by [calling the engine's `Process` method](https://github.com/onflow/flow-go/blob/d31fd63eb651ed9faf0f677e9934baef6c4d9792/network/p2p/network.go#L406), passing it the Flow ID of the message sender.
+
+[`Multicast`](https://github.com/onflow/flow-go/blob/4ddc17d1bee25c2ab12ceabcf814b702980fdebe/network/conduit.go#L82) is implemented by including a [`TargetIDs`](https://github.com/onflow/flow-go/blob/4ddc17d1bee25c2ab12ceabcf814b702980fdebe/network/message/message.proto#L12) field inside the message, which is published to a specific topic on the underlying gossip network. Upon receiving a new message on the gossip network, nodes must first [validate](https://github.com/onflow/flow-go/blob/4ddc17d1bee25c2ab12ceabcf814b702980fdebe/network/validator/targetValiator.go) that they are one of the intended recipients of the message before processing it.
+
+### Potential problems 
+
+The current network layer API was designed with the assumption that all messages sent and received either target or originate from staked Flow nodes. This is why an engine's [`Process`](https://github.com/onflow/flow-go/blob/master/network/engine.go#L28) method accepts a Flow ID identifying the message sender, and outgoing messages [must specify Flow ID(s)](https://github.com/onflow/flow-go/blob/master/network/conduit.go#L62) as targets.
+
+This assumption is no longer true today. The access node, for example, may communicate with multiple (unstaked) consensus followers. It's perceivable that in the future there will be even more cases where communication with unstaked parties may happen (for example, execution nodes talking to DPS).
+
+Currently, a [`Message`](https://github.com/onflow/flow-go/blob/698c77460bc33d1a8ee8a154f7fe4877bc518a02/network/message/message.proto) which is sent over the network contains many unnecessary fields which can be deduced by the receiver of the message. The only exceptions to this are the `Payload` field (which contains the actual message data) and the `TargetIDs` field (which is used by `Multicast`).
+
```

Comment:
Does this mean excluding the `OriginID` from the message? if that is the case, I would have [this concern](https://github.com/onflow/flow-go/pull/1167#discussion_r715231542) about breaking self-certifying feature of messages.

Reply:
Please see my response to that comment.

Reply:
Please see my response to that comment.

Reply:
A couple of things on top of that comment:

- besides the libp2p-based message signing, which is specifically a network-level whole-message signature by the source of the message, we do not have protocol message signing today,
- unless we mandate envelope signatures at the envelope level, which we currently do not do, the messages are not self-certifying through any feature of the protocol,
- we do have legitimate places where we have other signatures for components of the message (e.g. block proposal signatures from the block proposal emitter, or votes for a block within a QC), in general using the staking key. I do not believe that's because the staking keys are more authoritative.

Besides the fact that I fail to understand any guarantee that protocol-level message signing would bring us:
- adding message signatures (and verifying them) _using the networking key_ (on top of what already exists in libp2p) is an order of magnitude more expensive in (each of) CPU & bandwidth resources than any correspondence we may compute between a PeerID and a FlowID.
- adding message signatures (and verifying them) _using the staking key_ is an order of magnitude more expensive in CPU than doing the same thing with the networking key.

Reply:
Summary of my discussion with @yhassanzadeh13: 
- I think we're on the same page on this, but that the FLIP, as it is, fails to capture a nuance, let me try to explain:
- There is and should remain a space in the application for messages which application-level sender is not the application-level creator (and application-level signer) of the claimed artefact (including, but not limited to votes on a block, when retransmitted by the next views' leader)
- the FLIP should outline that this application-level indication of origination is legitimate, but should be handled within the application's message payload. 
- Indeed the point of the FLIP is that we're not removing an already-used claim to application-level origin.
- The fact of the matter is we're suggesting that in the following:
```
libp2p.msg { From, msg.Payload { OriginID, msg { <application-level protobuf> }, targetID? }, SIgningKey, Signature} 
```
From, for a valid message, is always equal (modulo format conversion) to OriginID. Hence the OriginID information is redundant. 
- But the use case for application-level origination is still valid. So we should also explain that, should we need one, we should (and do) encapsulate in the application data, in it this way:
```
libp2p.msg { From, msg.Payload { msg { <application-level part 1> { OriginID, sub-msg, OriginSignature}, <application-level part 2> }, targetID? }, SigningKey, Signature} 
```

Reply:
> Summary of my discussion with @yhassanzadeh13:
> 
> * I think we're on the same page on this, but that the FLIP, as it is, fails to capture a nuance, let me try to explain:
> * There is and should remain a space in the application for messages which application-level sender is not the application-level creator (and application-level signer) of the claimed artefact (including, but not limited to votes on a block, when retransmitted by the next views' leader)
> * the FLIP should outline that this application-level indication of origination is legitimate, but should be handled within the application's message payload.
> * Indeed the point of the FLIP is that we're not removing an already-used claim to application-level origin.
> * The fact of the matter is we're suggesting that in the following:
> 
> ```
> libp2p.msg { From, msg.Payload { OriginID, msg { <application-level protobuf> }, targetID? }, SIgningKey, Signature} 
> ```
> 
> From, for a valid message, is always equal (modulo format conversion) to OriginID. Hence the OriginID information is redundant.
> 
> * But the use case for application-level origination is still valid. So we should also explain that, should we need one, we should (and do) encapsulate in the application data, in it this way:
> 
> ```
> libp2p.msg { From, msg.Payload { msg { <application-level part 1> { OriginID, sub-msg, OriginSignature}, <application-level part 2> }, targetID? }, SigningKey, Signature} 
> ```

Thanks for summarizing the discussions @huitseeker, it perfectly addresses the concern I initially explained.